A case where @CalledMethods doesn't get all methods being called

I was running this checker in the gapic-generator (https://github.com/googleapis/gapic-generator). In PackageMetadataConfig.java:303 it reported warning as:

```
/gapic-generator/src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:303: error: [method.invocation.invalid] call to build() not allowed on the given receiver.
    return builder.build();
                        ^
  found   : @CalledMethods({"apiCommonVersionBound", "authVersionBound", "commonsCliVersionBound", "gaxGrpcVersionBound", "gaxHttpVersionBound", "gaxVersionBound", "generatedNonGAPackageVersionBound", "grpcVersionBound", "packageName", "protoVersionBound", "releaseLevel"}) Builder
  required: @CalledMethods({"apiCommonVersionBound", "apiVersion", "authVersionBound", "author", "commonsCliVersionBound", "email", "gaxGrpcVersionBound", "gaxHttpVersionBound", "gaxVersionBound", "generatedNonGAPackageVersionBound", "grpcVersionBound", "homepage", "licenseName", "packageName", "protoPackageDependencies", "protoPath", "protoVersionBound", "releaseLevel", "shortName"}) Builder
```

It indicates methods "author", "email", "homepage", "liscenseName", "protoPackageDependencies", "protoPath", "shortName" are not called by the builder which is not true refer to the code. 

```java
  public static PackageMetadataConfig createFromPackaging(
      ApiDefaultsConfig apiDefaultsConfig,
      DependenciesConfig dependenciesConfig,
      PackagingConfig packagingConfig) {
    // dependencies config
    Builder builder =
        newBuilder()
            .gaxVersionBound(dependenciesConfig.gaxVersionBound())
            .gaxGrpcVersionBound(dependenciesConfig.gaxGrpcVersionBound())
            .gaxHttpVersionBound(dependenciesConfig.gaxHttpVersionBound())
            .grpcVersionBound(dependenciesConfig.grpcVersionBound())
            .protoVersionBound(dependenciesConfig.protoVersionBound())
            .authVersionBound(dependenciesConfig.authVersionBound())
            .apiCommonVersionBound(dependenciesConfig.apiCommonVersionBound())
            .commonsCliVersionBound(dependenciesConfig.commonsCliVersionBound());

    // api_defaults config
    builder
 .generatedNonGAPackageVersionBound(apiDefaultsConfig.generatedNonGAPackageVersionBound())
        .author(apiDefaultsConfig.author())
        .email(apiDefaultsConfig.email())
        .homepage(apiDefaultsConfig.homepage())
        .licenseName(apiDefaultsConfig.licenseName());

    // packaging config
    builder.releaseLevel(apiDefaultsConfig.releaseLevel());
    if (packagingConfig != null) {
      builder
          .packageName(packagingConfig.packageName())
          .shortName(packagingConfig.apiName())
          .artifactType(packagingConfig.artifactType())
          .apiVersion(packagingConfig.apiVersion())
          .protoPath(packagingConfig.protoPath());

      // config depending on multiple sources
      Map<TargetLanguage, Map<String, VersionBound>> protoPackageDependencies =
          createProtoPackageDependencies(
              packagingConfig.protoPackageDependencies(), dependenciesConfig);
      builder.protoPackageDependencies(protoPackageDependencies);

      Map<TargetLanguage, Map<String, VersionBound>> protoPackageTestDependencies =
          createProtoPackageDependencies(
              packagingConfig.protoPackageTestDependencies(), dependenciesConfig);
      builder.protoPackageTestDependencies(protoPackageTestDependencies);
      if (packagingConfig.releaseLevel() != null) {
        builder.releaseLevel(packagingConfig.releaseLevel());
      }
    } else {
      builder
          .packageName("")
          .shortName("")
          .artifactType(PackagingArtifactType.GAPIC) // the most likely case
          .apiVersion("")
          .protoPath("")
          .protoPackageDependencies(ImmutableMap.of())
          .protoPackageTestDependencies(ImmutableMap.of());
    }

    return builder.build();
  }
```

--- a/typesafe-builder-checker/src/main/java/org/checkerframework/checker/builder/TypesafeBuilderAnnotatedTypeFactory.java
+++ b/typesafe-builder-checker/src/main/java/org/checkerframework/checker/builder/TypesafeBuilderAnnotatedTypeFactory.java
@@ -129,6 +129,21 @@ public class TypesafeBuilderAnnotatedTypeFactory extends BaseAnnotatedTypeFactor
   }
 
   /**
+   * Returns whether the return type of the given method invocation tree has an @This annotation
+   * from the Returns Receiver Checker.
+   *
+   * <p>Package-private to permit calls from TypesafeBuilderTransfer.
+   */
+  boolean returnsThis(final MethodInvocationTree tree) {
+    ReturnsRcvrAnnotatedTypeFactory rrATF = getReturnsRcvrAnnotatedTypeFactory();
+    ExecutableElement methodEle = TreeUtils.elementFromUse(tree);
+    AnnotatedTypeMirror methodATm = rrATF.getAnnotatedType(methodEle);
+    AnnotatedTypeMirror rrType =
+        ((AnnotatedTypeMirror.AnnotatedExecutableType) methodATm).getReturnType();
+    return rrType != null && rrType.hasAnnotation(This.class);
+  }
+
+  /**
    * This tree annotator is needed to create types for fluent builders that have @This annotations.
    */
   private class TypesafeBuilderTreeAnnotator extends TreeAnnotator {
@@ -142,14 +157,7 @@ public class TypesafeBuilderAnnotatedTypeFactory extends BaseAnnotatedTypeFactor
 
       // Check to see if the ReturnsReceiver Checker has a @This annotation
       // on the return type of the method
-
-      ReturnsRcvrAnnotatedTypeFactory rrATF = getReturnsRcvrAnnotatedTypeFactory();
-      ExecutableElement methodEle = TreeUtils.elementFromUse(tree);
-      AnnotatedTypeMirror methodATm = rrATF.getAnnotatedType(methodEle);
-      AnnotatedTypeMirror rrType =
-          ((AnnotatedTypeMirror.AnnotatedExecutableType) methodATm).getReturnType();
-
-      if (rrType != null && rrType.hasAnnotation(This.class)) {
+      if (returnsThis(tree)) {
 
         // Fetch the current type of the receiver, or top if none exists
         ExpressionTree receiverTree = TreeUtils.getReceiverTree(tree.getMethodSelect());

--- a/typesafe-builder-checker/src/main/java/org/checkerframework/checker/builder/TypesafeBuilderTransfer.java
+++ b/typesafe-builder-checker/src/main/java/org/checkerframework/checker/builder/TypesafeBuilderTransfer.java
@@ -1,5 +1,7 @@
 package org.checkerframework.checker.builder;
 
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
 import java.util.List;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -64,15 +66,42 @@ public class TypesafeBuilderTransfer extends CFTransfer {
 
     AnnotationMirror newType = atypefactory.createCalledMethods(newList.toArray(new String[0]));
 
-    Receiver receiverReceiver = FlowExpressions.internalReprOf(atypefactory, receiver);
-
     // For some reason, visitMethodInvocation returns a conditional store. I think this is to
     // support conditional post-condition annotations, based on the comments in CFAbstractTransfer.
     CFStore thenStore = result.getThenStore();
     CFStore elseStore = result.getElseStore();
 
-    thenStore.insertValue(receiverReceiver, newType);
-    elseStore.insertValue(receiverReceiver, newType);
+    while (receiver != null) {
+      // Insert the new type computed previously as the type of the receiver.
+      Receiver receiverReceiver = FlowExpressions.internalReprOf(atypefactory, receiver);
+      thenStore.insertValue(receiverReceiver, newType);
+      elseStore.insertValue(receiverReceiver, newType);
+
+      Tree receiverTree = receiver.getTree();
+
+      // Possibly recurse: if the receiver is itself a method call,
+      // then we need to also propagate this new information to its receiver
+      // if the method being called has an @This return type.
+      //
+      // Note that we must check for null, because the tree could be
+      // implicit (when calling an instance method on the class itself).
+      // In that case, do not attempt to refine either - the receiver is
+      // not a method invocation, anyway.
+      if (receiverTree == null || receiverTree.getKind() != Tree.Kind.METHOD_INVOCATION) {
+        // Do not continue, because the receiver isn't a method invocation itself. The
+        // end of the chain of calls has been reached.
+        break;
+      }
+
+      MethodInvocationTree receiverAsMethodInvocation = (MethodInvocationTree) receiver.getTree();
+
+      if (atypefactory.returnsThis(receiverAsMethodInvocation)) {
+        receiver = ((MethodInvocationNode) receiver).getTarget().getReceiver();
+      } else {
+        // Do not continue, because the method does not return @This.
+        break;
+      }
+    }
 
     return result;
   }

--- a/typesafe-builder-checker/tests/autovalue/Animal.java
+++ b/typesafe-builder-checker/tests/autovalue/Animal.java
@@ -71,4 +71,10 @@ abstract class Animal {
     Animal a1 = builder().setName("Jim").setNumberOfLegs(7).build();
     a1.toBuilder().build();
   }
+
+  public static void buildSomethingRightFluentWithLocal() {
+    Builder b = builder();
+    b.setName("Jim").setNumberOfLegs(7);
+    b.build();
+  }
 }

